src/pip/_internal/cli/base_command.py

```diff
@@ -186,6 +188,18 @@ def _main(self, args):
                 "This will become an error in pip 21.0."
             )
 
+        runner = self.run if self._is_debug() else self._trap_run
+        status = runner(options, args)
+        self.handle_pip_version_check(options)
+        return status
+
+    @staticmethod
+    def _is_debug():
+        return any(
```

Comment:
I measured the cost of this approach. In my environment, it had to iterate over as many as 34k objects and added ~100ms to the startup time. If this approach is the one we would prefer, it probably would be worthwhile to add some instrumentation to `bdb.Bdb` to keep track of instances of it.

Reply:
Uhm, I'm not a big fan of this. I'd much rather that we don't fix this that to add a 100ms cost to every pip execution.

Reply:
I agree.

Reply:
I had a thought - what about combining the imperfect "if pdb/bdb in sys.modules" with this code. That would short-circuit early the most common scenario, but allow the more expensive operation to continue.

Reply:
In b82f628e6, I add that check and now the startup is fast in the common case.